torch/csrc/jit/codegen/cuda/lower_unroll.cpp

```diff
@@ -32,6 +32,221 @@ kir::ForLoop* cloneLoopNest(const kir::ForLoop* for_loop, bool unroll = false) {
   return new_loop;
 }
 
+// Create a new vectorize For-Loop
+// Add For-Loop to If-Then-Else parent scope
+// for (index = start; index < extent; index += offset)
+// vectorize flag - Do not generate for-loop
+// shift value - Add shift to global indices generated within For-Loop
+void cloneVectorizeLoopNests(
+    kir::IfThenElse* parent_ite,
+    const std::vector<kir::ForLoop*>& for_loops,
+    kir::Val* extent,
+    bool vectorize,
+    kir::Val* shift) {
+  kir::IrBuilder ir_builder(GpuLower::current()->kernel());
+
+  for (auto fl : for_loops) {
+    auto first_expr = fl->body().exprs().front();
+    bool has_vectorize_op =
+        (first_expr->isA<kir::UnaryOp>() &&
+         first_expr->as<kir::UnaryOp>()->operation() == UnaryOpType::Set);
+
+    TORCH_INTERNAL_ASSERT(!has_vectorize_op || fl->body().exprs().size() == 1);
+
+    const auto new_loop = ir_builder.create<kir::ForLoop>(
+        fl->index(),
+        extent,
+        fl->iter_domain(),
+        false,
+        vectorize && has_vectorize_op,
+        shift);
+
+    for (auto expr : fl->body().exprs()) {
+      new_loop->body().push_back(expr);
+    }
+
+    parent_ite->thenBody().push_back(new_loop);
+  }
+}
+
+// Find any child For-Loops
+// Add remaining expressions to new parent For-Loop
+std::vector<kir::ForLoop*> parseVectorizedForLoop(
+    const kir::ForLoop* for_loop,
+    kir::ForLoop* new_loop) {
+  std::vector<kir::ForLoop*> loops;
+  for (auto expr : for_loop->body().exprs()) {
+    if (auto nested_for_loop = dynamic_cast<kir::ForLoop*>(expr)) {
+      loops.push_back(nested_for_loop);
+    } else {
+      new_loop->body().push_back(expr);
+    }
+  }
+  return loops;
+}
+
+// Find the first vectorize set - either read or write
+// Add child For-Loop to loop_structure
+// Enable vectorize flag in child For-Loop
+kir::Expr* findVectorizedSet(
+    std::vector<kir::ForLoop*>& loop_structure,
+    const std::vector<kir::ForLoop*>& for_loops) {
+  for (auto fl : for_loops) {
+    auto first_expr = fl->body().exprs().front();
+    bool has_vectorize_op =
```

Comment:
Does this mean that vectorization only happens the first expression is a UnaryOp::Set? What happens if other expressions exist in the same loop?

Reply:
No. We are using the first vectorized set op to get the vectorized index. The assumption is that the shift offset is the same for all the global tensors. We can check this assumption in `validateVectorizedTensors`.

Reply:
I don't understand this assumption, can you elaborate?